// Authors: Chen Qian(qcdsr970209@gmail.com)

syntax = "proto2";

option cc_generic_services = true;

package ctgfs;

message ClientKVRequest {
    required uint64 id = 1;
    required string addr = 2;

    message Command {
        enum Type {
            kCreateDir = 0;
            kRemoveDir = 1;
            kWriteFile = 2;
            kRemoveFile = 3;
            kReadFile = 4;
        }
        required Type type = 2;
        required string path = 3;
        optional string value = 4;
    }
    optional Command command = 3;
};

message ClientKVResponse {
	required uint64 id = 1;
	required string addr = 2;
};


// send the pacage at regular time
// shared the info from filesystem
// you should give the info to master so that master can give choice
message HeartBeatMessageRequest {
  enum HeartBeatType {
    kRegist = 0;
    kInfoUpdate = 1;
    kUnregist = 2;
  }
  required HeartBeatType type = 1;
  // ip:port
  // Example : 
  // 127.0.0.1:1234
  required string addr = 2;
  // should add more info later

  // the information of the file system.
  required uint32 file_num = 3;
  required uint32 disk_usage = 4;

};

message HeartBeatMessageResponse { 
  // inform the file system the heart beat has been accecpted 
  // it can add more info later
};

// master service
service MasterService{
	rpc ClientAskForKV(ClientKVRequest) returns (ClientKVResponse);	
  rpc SendHeartBeat(HeartBeatMessageRequest) returns (HeartBeatMessageResponse);
};
